iocore/net/quic/QUICPacketProtectionKeyInfo.h

```diff
@@ -0,0 +1,123 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#pragma once
+
+#include "QUICTypes.h"
+#include "QUICKeyGenerator.h"
+
+class QUICPacketProtectionKeyInfo
+{
+public:
+  enum class Context { SERVER, CLIENT };
+
+  // FIXME This should be passed to the constructor but NetVC cannot pass it because it has set_context too.
+  void set_context(Context ctx);
+
+  void drop_keys(QUICKeyPhase phase);
+
+  // Payload Protection (common)
+
+  virtual const QUIC_EVP_CIPHER *get_cipher(QUICKeyPhase phase) const;
+  virtual size_t get_tag_len(QUICKeyPhase phase) const;
+  virtual void set_cipher_initial(const QUIC_EVP_CIPHER *cipher);
+  virtual void set_cipher(const QUIC_EVP_CIPHER *cipher, size_t tag_len);
+
+  // Payload Protection (encryption)
+
+  virtual bool is_encryption_key_available(QUICKeyPhase phase) const;
+  virtual void set_encryption_key_available(QUICKeyPhase phase);
+
+  virtual const uint8_t *encryption_key(QUICKeyPhase phase) const;
+  virtual uint8_t *encryption_key(QUICKeyPhase phase);
+
+  virtual size_t encryption_key_len(QUICKeyPhase phase) const;
+
+  virtual const uint8_t *encryption_iv(QUICKeyPhase phase) const;
+  virtual uint8_t *encryption_iv(QUICKeyPhase phase);
+
+  virtual const size_t *encryption_iv_len(QUICKeyPhase phase) const;
+  virtual size_t *encryption_iv_len(QUICKeyPhase phase);
+
+  // Payload Protection (decryption)
+
+  bool is_decryption_key_available(QUICKeyPhase phase) const;
+  void set_decryption_key_available(QUICKeyPhase phase);
+
+  const uint8_t *decryption_key(QUICKeyPhase phase) const;
+  uint8_t *decryption_key(QUICKeyPhase phase);
+
+  size_t decryption_key_len(QUICKeyPhase phase) const;
+
+  const uint8_t *decryption_iv(QUICKeyPhase phase) const;
+  uint8_t *decryption_iv(QUICKeyPhase phase);
+
+  const size_t *decryption_iv_len(QUICKeyPhase phase) const;
+  size_t *decryption_iv_len(QUICKeyPhase phase);
+
+  // Header Protection
+
+  const QUIC_EVP_CIPHER *get_cipher_for_hp(QUICKeyPhase phase) const;
+  void set_cipher_for_hp_initial(const QUIC_EVP_CIPHER *cipher);
+  void set_cipher_for_hp(const QUIC_EVP_CIPHER *cipher);
+
+  const uint8_t *encryption_key_for_hp(QUICKeyPhase phase) const;
+  uint8_t *encryption_key_for_hp(QUICKeyPhase phase);
+
+  size_t encryption_key_for_hp_len(QUICKeyPhase phase) const;
+
+  const uint8_t *decryption_key_for_hp(QUICKeyPhase phase) const;
+  uint8_t *decryption_key_for_hp(QUICKeyPhase phase);
+
+  size_t decryption_key_for_hp_len(QUICKeyPhase phase) const;
+
+private:
+  Context _ctx = Context::SERVER;
+
+  // Payload Protection
+
+  const QUIC_EVP_CIPHER *_cipher_initial = nullptr;
+  const QUIC_EVP_CIPHER *_cipher         = nullptr;
+  size_t _tag_len                        = 0;
+
+  bool _is_client_key_available[5] = {false};
+  bool _is_server_key_available[5] = {false};
+
+  // FIXME EVP_MAX_KEY_LENGTH and EVP_MAX_IV_LENGTH are not enough somehow
+  uint8_t _client_key[5][512];
```

Comment:
Just curious about why array of KeyMaterial doesn't work.

Reply:
Not sure :) These are came from the old code, and IIRC, it seemed like buffer overflow.